src/indexer.rs

```diff
@@ -0,0 +1,624 @@
+    /// node to avoid overloading the node. Defaults to 1 second.
+    pub fn set_wait_after_failure(self, wait_after_fail: Duration) -> Self {
+        Self {
+            wait_after_fail,
+            ..self
+        }
+    }
+
+    /// Add an additional endpoint to the list of endpoints the indexer will
+    /// use. This is added to the end of the list, so the endpoint is only tried
+    /// in case of failure of previous queries.
+    pub fn push_endpoint(mut self, endpoint: v2::Endpoint) -> Self {
+        self.endpoints.push(endpoint);
+        self
+    }
+
+    /// Set the maximum number of blocks that will be queried in parallel, if
+    /// they are available. Defaults to 4 if not set explicitly.
+    pub fn set_max_parallel(self, max_parallel: usize) -> Self {
+        Self {
+            max_parallel,
+            ..self
+        }
+    }
+
+    /// Traverse the chain according to the supplied configuration, invoking
+    /// [`on_finalized`](Indexer::on_finalized) for each finalized block.
+    ///
+    /// Multiple [`on_finalized`](Indexer::on_finalized) calls might be executed
+    /// concurrently, but their responses will be written to the provided
+    /// [`tokio::sync::mpsc::Sender`] in the increasing order of block height,
+    /// with no gaps.
+    ///
+    /// If a query fails, either due to timeout or node not being available the
+    /// indexer will attempt the next endpoint it is configured with.
+    ///
+    /// For robust indexing it is crucial that the supplied
+    /// [`Endpoints`](v2::Endpoint) are configured with timeouts so that the
+    /// indexer may make progress.
+    ///
+    /// The [`traverse`](Self::traverse) method will return either when
+    /// signalled so by the [`on_failure`](Indexer::on_failure) method or
+    /// when the receiver part of the supplied [`tokio::sync::mpsc::Sender`]
+    /// is closed. Typically this method should run in a task spawned via
+    /// [`tokio::spawn`].
+    pub async fn traverse<I: Indexer>(
+        self,
+        mut indexer: I,
+        sender: tokio::sync::mpsc::Sender<I::Data>,
+    ) -> QueryResult<()> {
+        let TraverseConfig {
+            endpoints,
+            max_parallel,
+            max_behind,
+            wait_after_fail,
+            start_height: mut height,
+        } = self;
+        let mut successive_failures: u64 = 0;
+        for node_ep in endpoints.into_iter().cycle() {
+            if sender.is_closed() {
+                return Ok(());
+            }
+            if successive_failures > 0 {
+                tokio::time::sleep(wait_after_fail).await
+            }
+            let mut node = match v2::Client::new(node_ep.clone()).await {
+                Ok(v) => v,
+                Err(e) => {
+                    successive_failures += 1;
+                    let should_stop = indexer
+                        .on_failure(
+                            node_ep,
+                            successive_failures + 1,
```

Comment:
**Change request:** I think this is reporting a `successive_failures` which is off by one, since it was already incremented before the `on_failure` hook.
```suggestion
                    successive_failures += 1;
                    let should_stop = indexer
                        .on_failure(
                            node_ep,
                            successive_failures,
```
Likewise with all the other calls to this hook.